File: LeetCode/Problems/C++/17. Letter Combinations of a Phone Number.cpp
class Solution {
public:
    void backtracking(vector<string>& result, unordered_map<char,vector<char>>& lookup, string& cur, string digits, int idx){
        if(idx==digits.size()) {
            result.push_back(cur);
            return ;
        }
        vector<char> candid = lookup[digits[idx]];
        for(int i=0; i<candid.size();i++){
            cur[idx] = candid[i];
            backtracking(result, lookup, cur, digits, idx+1);
        }

        
    }
    vector<string> letterCombinations(string digits) {
        if(!digits.size()) return {};
        vector<string> result;
        unordered_map<char,vector<char>> lookup;
        
        lookup['2'] = vector<char> {'a', 'b', 'c'};
        lookup['3'] = vector<char> {'d', 'e', 'f'};
        lookup['4'] = vector<char> {'g', 'h', 'i'};
        lookup['5'] = vector<char> {'j', 'k', 'l'};
        lookup['6'] = vector<char> {'m', 'n', 'o'};
        lookup['7'] = vector<char> {'p', 'q', 'r', 's'};
        lookup['8'] = vector<char> {'t', 'u', 'v'};
        lookup['9'] = vector<char> {'w', 'x', 'y', 'z'};
        string cur = "";
        for(int i=0;i<digits.size();i++) cur.append(" ");
        
        backtracking(result, lookup, cur, digits, 0);
        return result;
    }
};

File: LeetCode/Problems/Python/17. Letter Combinations of a Phone Number.py
#baccktracking
#Time: O(3^m*4^n)
#Space: O(3^m*4^n)
class Solution:
    def letterCombination(self, digits, dict, result, curr):
        if not digits:
            result.append(curr)
            return
        
        for ch in dict[digits[0]]:
            self.letterCombination(digits[1:], dict, result, curr+ch)
        
    def letterCombinations(self, digits: str) -> List[str]:
        if not digits:
            return []

        dict =  {'2': ['a', 'b', 'c'], '3': ['d', 'e', 'f'], '4': ['g', 'h', 'i'], '5': ['j', 'k', 'l'], '6': ['m', 'n', 'o'], '7': ['p', 'q', 'r', 's'], '8': ['t', 'u', 'v'], '9': ['w', 'x', 'y', 'z']}


        result = []
        self.letterCombination(digits, dict, result, "")
       
        return result